### PR TITLE
Milestone 1 Task 1: configure workflow persistence

### PR DESCRIPTION
**Pull request by ShaojieJiang**
## Summary
- add configuration helpers to load checkpoint and hosting settings from the environment
- introduce a persistence factory that provisions SQLite or Postgres LangGraph checkpointers and update the FastAPI server to use it
- document the milestone 1 architecture decisions and mark task 1 complete in the roadmap

## Testing
- uv run pytest -q

------
https://chatgpt.com/codex/tasks/task_e_68d9296b30888327a0e2334a59f18806

### CODE REVIEW

**Comment by ShaojieJiang** [comment]
Fix the failed checks and make sure the unit test coverage is 100% @codex

**Comment by ShaojieJiang** [comment]
There are still CI failures. Fix them @codex

**Comment by ShaojieJiang** [comment]
Can you fix the linting errors? @claude

File: src/orcheo/config.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+"""Runtime configuration helpers for Orcheo."""
```
Comment (ShaojieJiang) [inline src/orcheo/config.py:1]: @codex @claude do you think there are better approaches to managing runtime configurations? Relying on external dependencies is fine.

Comment (ShaojieJiang) [reply src/orcheo/config.py:1]: @claude Change the PR to use Dynaconf

**Comment by ShaojieJiang** [comment]
@codex @claude, review the latest code

**Comment by ShaojieJiang** [comment]
> **Claude finished @ShaojieJiang's task** —— [View job](https://github.com/ShaojieJiang/aic-flow/actions/runs/18137761479)
> 
> ### Code Review for Milestone 1 Task 1: Workflow Persistence
> * [x]  Read and analyze all changed files
> * [x]  Review configuration management implementation
> * [x]  Review persistence factory implementation
> * [x]  Review test coverage and quality
> * [x]  Check documentation and architecture decisions
> * [x]  Provide comprehensive review feedback
> 
> ## Overall Assessment
> This is a solid implementation of the workflow persistence infrastructure. The architecture decisions are well-documented and the code demonstrates good separation of concerns. The migration to Dynaconf was a good choice for configuration management.
> 
> ## Key Strengths
> ✅ **Clean Architecture**: Clear separation between config, persistence, and main application ✅ **Comprehensive Testing**: 100% test coverage with edge cases covered ✅ **Good Documentation**: Excellent milestone documentation explaining decisions ✅ **Type Safety**: Proper type hints and validation throughout ✅ **Error Handling**: Descriptive error messages for configuration issues
> 
> ## Detailed Review
> ### Configuration Management (`src/orcheo/config.py`)
> **Strengths:**
> 
> * Good use of Dynaconf for environment variable management
> * Proper validation with descriptive error messages
> * Type-safe configuration with `CheckpointBackend` literal type
> * Effective caching with `@lru_cache` and refresh capability
> 
> **Areas for Improvement:**
> 
> * **Line 33**: The `str(backend_raw or _DEFAULTS["CHECKPOINT_BACKEND"]).lower()` logic is defensive but could be clearer. Consider splitting into separate checks:
>   ```python
>   backend_raw = source.get("CHECKPOINT_BACKEND", _DEFAULTS["CHECKPOINT_BACKEND"])
>   if backend_raw is None:
>       backend = str(_DEFAULTS["CHECKPOINT_BACKEND"]).lower()
>   else:
>       backend = str(backend_raw).lower()
>   ```
> 
> ### Persistence Layer (`src/orcheo/persistence.py`)
> **Strengths:**
> 
> * Excellent use of context manager pattern for resource management
> * Proper handling of optional PostgreSQL dependencies
> * Clear separation of SQLite and PostgreSQL backends
> 
> **Critical Issue:**
> 
> * **Lines 54-59**: There's a logical error in the PostgreSQL connection handling. The code creates a connection pool but doesn't open it before trying to borrow a connection:
>   ```python
>   pool = AsyncConnectionPool(dsn)
>   try:
>       async with pool.connection() as conn:  # This will fail
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   **Fix needed:**
>   ```python
>   pool = AsyncConnectionPool(dsn)
>   await pool.open()  # Add this line
>   try:
>       async with pool.connection() as conn:
>           yield AsyncPostgresSaver(cast(Any, conn))
>   finally:
>       await pool.close()
>   ```
> 
> **Minor Issues:**
> 
> * **Line 56**: The `# type: ignore[assignment]` comment could be more specific about why the type ignore is needed
> 
> ### Main Application (`src/orcheo/main.py`)
> **Strengths:**
> 
> * Clean integration of the persistence layer
> * Proper use of async context managers
> * Good error handling and logging
> 
> **Areas for Improvement:**
> 
> * **Lines 23**: Global `settings = get_settings()` call could cause issues if environment changes during runtime. Consider moving this inside functions that need it
> * **Line 52**: The persistence context manager is properly used within the workflow execution
> 
> ### Test Quality (`tests/`)
> **Excellent test coverage** with several standout practices:
> 
> **`test_config.py`:**
> 
> * Good test isolation with `monkeypatch`
> * Tests both positive and negative cases
> * Validates refresh behavior
> 
> **`test_persistence.py`:**
> 
> * Excellent mocking strategy for external dependencies
> * Tests both SQLite and PostgreSQL backends
> * Covers error cases with invalid backends
> 
> **`test_main.py`:**
> 
> * Good use of context managers for mocking
> * Tests the integration between components
> 
> **Minor suggestion:** Consider adding integration tests that actually create real SQLite connections to validate the complete flow.
> 
> ### Documentation (`docs/milestone1_task1.md`)
> **Outstanding documentation** that clearly explains:
> 
> * Architecture decisions and rationale
> * Technology choices (SQLite vs PostgreSQL)
> * Configuration approach
> * Future considerations
> 
> ## Issues to Address
> ### High Priority
> 1. **PostgreSQL Connection Pool**: Fix the pool opening issue in `persistence.py:54-59`
> 
> ### Medium Priority
> 2. **Global Settings**: Consider moving the global settings call inside functions for better testability
> 3. **Type Ignore**: Add more specific type ignore comments
> 
> ### Low Priority
> 4. **Configuration Logic**: Simplify the backend validation logic for better readability
> 
> ## Security & Best Practices
> ✅ **No hardcoded secrets or credentials** ✅ **Proper environment variable handling** ✅ **Safe database connection management** ✅ **Good error boundaries**
> 
> ## Performance Considerations
> * The `@lru_cache` on settings is appropriate for this use case
> * PostgreSQL connection pooling is correctly implemented (once the opening issue is fixed)
> * Async context managers ensure proper resource cleanup
> 
> ## Conclusion
> This is a well-architected solution that successfully addresses Milestone 1 Task 1 requirements. The main issue is the PostgreSQL pool initialization bug, which should be fixed before merging. The code demonstrates good engineering practices and will provide a solid foundation for the workflow orchestration platform.
> 
> **Recommendation:** Fix the PostgreSQL pool issue and this will be ready to merge.

@claude apply the suggested changes
